src/wizeng.main.v3

```diff
@@ -35,239 +36,60 @@ def main(args: Array<string>) -> int {
 		engine.extensions = EngineOptions.extensions;
 		pos++;
 	}
-	// Find range of all .wasm files
+	// Find range of all .wasm or .bin.wast files
 	var before = pos;
 	var skip = 0;
 	while (pos < args.length) {
 		var a = args[pos];
 		if (a.length == 0) break;
 		if (Strings.equal("--", a)) { skip = 1; break; } // skip argument separator
-		if (!Strings.endsWith(a, ".wasm")) break;
+		if (!Strings.endsWith(a, ".wasm") && !Strings.endsWith(a, ".bin.wast")) break;
 		pos++;
 	}
-	wasmFiles = args[before ... pos];
+	for (i = before; i < pos; i++) {
+		if (Strings.endsWith(args[i], ".wasm")) wasm.put(args[i]);
+		else wast.put(args[i]);
+	}
 	// Program arguments remain after .wasm files or after "--"
 	var v = Vector<string>.new();
-	if (wasmFiles.length > 0) v.put(wasmFiles[wasmFiles.length - 1]);
+	if (wasm.length > 0) v.put(wasm[wasm.length - 1]);
 	v.putr(args[(pos + skip) ...]);
 	var programArgs = v.extract();
 
 	if (WizengOptions.PREGEN_FILE.val != null) {
 		var error = Target.pregenIntoFile(WizengOptions.PREGEN_FILE.val);
 		if (error != null) return error.exit(2);
-		if (wasmFiles.length == 0) return 0; // all done
+		if (wasm.length == 0) return 0; // all done
 	}
 
 	if (WizengOptions.helped) return 0;  // printed help, nothing more to do
 
-	if (wasmFiles.length == 0) return ErrorBuilder.new()
+	if (wasm.length == 0 && wast.length == 0) return ErrorBuilder.new()
 		.puts("wizeng: no input files\n")
 		.exit(3);
 
 	// Expose the wizeng HostModule if it's an activated extension
 	if (engine.extensions.WIZENG) ModuleRegistry.expose("wizeng");
 
-	for (i < wasmFiles.length) {
-		var path = wasmFiles[i];
-		if (!Strings.endsWith(path, ".wasm")) return ErrorBuilder.new()
-			.puts(path)
-			.puts(": unknown file format, expected .wasm\n")
-			.exit(4);
-	}
 	var monitors = MonitorOptions.getMonitors().extract();
-	var hasMonitors = monitors.length > 0 || BasicTracing.requiresInstrumentation();
-
-	if (hasMonitors) Execute.tiering.onMonitorsStart();
-
-	// Load an instrument all modules.
-	var t = loadAndInstrumentModules(engine, monitors, wasmFiles);
-	if (t.0 != null) return ErrorBuilder.new()
-		.render(t.0.error_code, t.0.filename, t.0.section, t.0.error_index, t.0.error_pos, t.0.error_msg)
-		.exit(3);
-
-	// Loading succeeded. Link modules and execute (potentially repeatedly).
+	var wasmFiles = wasm.extract();
+	var wastFiles = wast.extract();
 	var exit_code = 0;
-	for (i < WizengOptions.REPEAT.val) {
-		exit_code = linkAndExecuteMain(engine, monitors, t.1, programArgs);
-		if (exit_code != 0) break;
+	
+	if (wasmFiles.length > 0) {
+		exit_code = WasmMode.run(engine, monitors, wasmFiles, programArgs);
+		if (exit_code != 0) return exit_code;
+	}
+	if (wastFiles.length > 0) {
+		exit_code = SpectestMode.run(engine, monitors, wastFiles);
```

Comment:
It'd be good to have a build option `--no-spec-test` which doesn't include the spec test runner at all (and maybe similarly `--no-wasm-run`).

One way to support that is to have a `var` field in this component that is a function for running spec tests, and have the `SpecTestMode` initialize that field to point to one of its functions (at compile time). That way, if the `SpecTestMode` code is not included at build time, the `var` field will be null; the `main` runner can then be adjusted accordingly.